File: java/03_numero_espanol/SpanishNumber.java
import java.util.Map;

public class SpanishNumber {
    public static String numeroAEspanol(int n) {
        String[] u = {"cero","uno","dos","tres","cuatro","cinco","seis","siete","ocho","nueve"};
        Map<Integer,String> e = Map.ofEntries(Map.entry(10,"diez"),Map.entry(11,"once"),Map.entry(12,"doce"),Map.entry(13,"trece"),Map.entry(14,"catorce"),Map.entry(15,"quince"),Map.entry(16,"dieciséis"),Map.entry(17,"diecisiete"),Map.entry(18,"dieciocho"),Map.entry(19,"diecinueve"),Map.entry(20,"veinte"),Map.entry(21,"veintiuno"),Map.entry(22,"veintidós"),Map.entry(23,"veintitrés"),Map.entry(24,"veinticuatro"),Map.entry(25,"veinticinco"),Map.entry(26,"veintiséis"),Map.entry(27,"veintisiete"),Map.entry(28,"veintiocho"),Map.entry(29,"veintinueve"));
        Map<Integer,String> d = Map.of(30,"treinta",40,"cuarenta",50,"cincuenta",60,"sesenta",70,"setenta",80,"ochenta",90,"noventa");
        Map<Integer,String> c = Map.of(100,"cien",200,"doscientos",300,"trescientos",400,"cuatrocientos",500,"quinientos",600,"seiscientos",700,"setecientos",800,"ochocientos",900,"novecientos");
        if (n < 0) return "menos " + numeroAEspanol(Math.abs(n));
        if (n < 10) return u[n];
        if (n <= 29) return e.get(n);
        if (n < 100) { int dec = n/10*10, uni = n%10; return uni == 0 ? d.get(dec) : d.get(dec) + " y " + u[uni]; }
        if (n == 100) return "cien";
        if (n < 1000) { int cen = n/100*100, res = n%100; String base = cen == 100 ? "ciento" : c.get(cen); return res == 0 ? c.get(cen) : base + " " + numeroAEspanol(res); }
        if (n < 1000000) { int miles = n/1000, resto = n%1000; String texto = miles == 1 ? "mil" : numeroAEspanol(miles) + " mil"; return resto == 0 ? texto : texto + " " + numeroAEspanol(resto); }
        return "Número fuera de rango";
    }
}

File: perl/03_numero_espanol/spanish_number.pl
sub numero_a_espanol {
    my ($n) = @_;
    my @u = qw(cero uno dos tres cuatro cinco seis siete ocho nueve);
    my %e = (10=>"diez",11=>"once",12=>"doce",13=>"trece",14=>"catorce",15=>"quince",16=>"dieciséis",17=>"diecisiete",18=>"dieciocho",19=>"diecinueve",20=>"veinte",21=>"veintiuno",22=>"veintidós",23=>"veintitrés",24=>"veinticuatro",25=>"veinticinco",26=>"veintiséis",27=>"veintisiete",28=>"veintiocho",29=>"veintinueve");
    my %d = (30=>"treinta",40=>"cuarenta",50=>"cincuenta",60=>"sesenta",70=>"setenta",80=>"ochenta",90=>"noventa");
    my %c = (100=>"cien",200=>"doscientos",300=>"trescientos",400=>"cuatrocientos",500=>"quinientos",600=>"seiscientos",700=>"setecientos",800=>"ochocientos",900=>"novecientos");
    return "menos " . numero_a_espanol(abs($n)) if $n < 0;
    return $u[$n] if $n < 10;
    return $e{$n} if $n <= 29;
    if ($n < 100) { my $dec = int($n/10)*10; my $uni = $n%10; return $uni == 0 ? $d{$dec} : "$d{$dec} y $u[$uni]"; }
    return "cien" if $n == 100;
    if ($n < 1000) { my $cen = int($n/100)*100; my $res = $n%100; my $base = $cen == 100 ? "ciento" : $c{$cen}; return $res == 0 ? $c{$cen} : "$base " . numero_a_espanol($res); }
    if ($n < 1000000) { my $miles = int($n/1000); my $resto = $n%1000; my $texto = $miles == 1 ? "mil" : numero_a_espanol($miles) . " mil"; return $resto == 0 ? $texto : "$texto " . numero_a_espanol($resto); }
    return "Número fuera de rango";
}
1;
